rgs.1

.\" Process this file with
.\" groff -man -Tascii foo.1
.\"
.TH rgs 1 2020-2023 Linux "lg archive"
.SH NAME
rgs - a shell command to manipulate a remote SBC's GPIO.

.SH SYNOPSIS

.B rgpiod &

then

.B rgs {command}+

.SH DESCRIPTION

.ad l

.nh


.br
rgs is a program which allows remote control of the GPIO and
other functions of Linux SBCs running the rgpiod daemon.

.br
The rgpiod daemon must be running on the SBCs you wish to control.

.br
.SS Features
.br

.br
o reading and writing GPIO singly and in groups

.br
o software timed PWM and waves

.br
o GPIO callbacks

.br
o pipe notification of GPIO events

.br
o I2C wrapper

.br
o SPI wrapper

.br
o serial link wrapper

.br
o simple file handling

.br
o creating and running scripts on the rgpiod daemon

.br
.SS Usage
.br
rgs {command}+

.br
rgs will show the result of the command on screen.

.br
The rgs process returns an exit status (which can be displayed with
the command echo $?).

.br

.EX
RGS_OK            0
.br
RGS_CONNECT_ERR 255
.br
RGS_OPTION_ERR  254
.br
RGS_SCRIPT_ERR  253
.br
.br
.br

.EE

.br
If an error was detected a message will have been written to stderr.
This is likely to be more informative than the message returned by rgs.

.br
Several commands may be entered on a line.  If present PROC and PARSE must
be the last command on a line.

.br
.SS Notes
.br
rgs does not show the status of successful commands unless the
command itself returns data.  The status (0) will be returned to
rgs but will be discarded.

.br
When a command takes a number as a parameter it may be entered as hex
(precede by 0x), octal (precede by 0), or decimal.

.br
E.g. 23 is 23 decimal, 0x100 is 256 decimal, 070 is 56 decimal.

.br
Some commands can return a variable number of data bytes.  By
default this data is displayed as decimal.  The rgs -a option
can be used to force the display as ASCII and the rgs -x
option can be used to force the display as hex.

.br
E.g. assuming the transmitted serial data is the letters ABCDEONM

.br

.EX
$ rgs serr 4 100 # assumes serial data available from handle 4
.br
8 65 66 67 68 69 79 78 77
.br

.br
$ rgs -a serr 4 100
.br
8 ABCDEONM
.br

.br
$ rgs -x serr 4 100
.br
8 41 42 43 44 45 4f 4e 4d
.br

.EE

.br
.SS Permissions
.br
Generally objects created on the rgpiod daemon exist for the duration of the
socket connection.

.br
For a Python script this will be for the duration of the script.  For a
program linked with rgpio this will be for the duration of the program.

.br
For rgs it is the command line.

.br
This means that the following command will achieve little

.br

.EX
rgs go 0 # get handle to gpiochip 0
.br

.EE

.br
The daemon will delete the handle as soon as the rgs command has finished.

.br
To preserve the handle it must be shared.

.br
A lot of the examples will show the command c 1 (use share id 1).
This means the handle is preserved and may be used in subsequent commands.

.br

.EX
rgs c 1 go 0 # get and preserve handle to gpiochip 0
.br

.EE

.br
If the LG_SHARE environment variable exists that share will be
automatically used in rgs commands.

.br
E.g. \fBexport LG_SHARE=12\fP will automatically use share id 12.

.br
If a command is privileged it is indicated in the notes for the
command.  The examples given here assume the daemon access control
system is not active (so any user can use privileged commands).

.br
If the LG_USER environment variable exists that user will be
automatically used in rgs commands.  This only has an effect if
the rgpiod daemon is running with access control enabled.

.br
E.g. \fBexport LG_USER=joan\fP will automatically use user joan.

.br

.SH OVERVIEW
.SS FILES
.B FO file mode 
File open
.P
.B FC h 
File close
.P
.B FR h num 
File read
.P
.B FW h bvs 
File write
.P
.B FS h num from 
File seek
.P
.B FL pat num 
File list
.P
.SS GPIO
.br
.B GO gc 
gpiochip open device
.P
.B GC h 
gpiochip close device
.P
.B GIC h 
gpiochip information
.P
.B GIL h g 
gpiochip line information
.P
.B GMODE h g 
GPIO get mode
.P
.B GSI h g 
GPIO claim for input (simple)
.P
.B GSIX h lf g 
GPIO claim for input
.P
.B GSO h g 
GPIO claim for output (simple)
.P
.B GSOX h lf g v 
GPIO claim for output
.P
.B GSA h g nfyh 
GPIO claim for alerts (simple)
.P
.B GSAX h lf ef g nfyh 
GPIO claim for alerts
.P
.B GSF h g 
GPIO free
.P
.B GSGI h g* 
GPIO group claim for inputs (simple)
.P
.B GSGIX h lf g* 
GPIO group claim for inputs
.P
.B GSGO h g* 
GPIO group claim for outputs (simple)
.P
.B GSGOX h lf g* v* 
GPIO group claim for outputs
.P
.B GSGF h g 
GPIO group free
.P
.B GR h g 
GPIO read
.P
.B GW h g v 
GPIO write
.P
.B GGR h g 
GPIO group read
.P
.B GGW h g gbits 
GPIO group write (simple)
.P
.B GGWX h g gbits gmask 
GPIO group write
.P
.B GP h g mon moff 
GPIO tx pulse (simple)
.P
.B GPX h g mon moff off cyc 
GPIO tx pulse
.P
.B P h g pf pdc 
GPIO tx PWM (simple)
.P
.B PX h g pf pdc off cyc 
GPIO tx PWM
.P
.B S h g spw 
GPIO tx servo pulses (simple)
.P
.B SX h g spw sf off cyc 
GPIO tx servo pulses
.P
.B GWAVE h g p* 
GPIO group tx wave
.P
.B GBUSY h g k 
GPIO or group tx busy
.P
.B GROOM h g k 
GPIO or group tx entries
.P
.B GDEB h g us 
GPIO debounce time
.P
.B GWDOG h g us 
GPIO watchdog time
.P
.SS I2C
.B I2CO ib id if 
I2C open device
.P
.B I2CC h 
I2C close device
.P
.B I2CWQ h bit 
SMB Write Quick: write bit
.P
.B I2CRS h 
SMB Read Byte: read byte
.P
.B I2CWS h bv 
SMB Write Byte: write byte
.P
.B I2CRB h r 
SMB Read Byte Data: read byte from register
.P
.B I2CWB h r bv 
SMB Write Byte Data: write byte to register
.P
.B I2CRW h r 
SMB Read Word Data: read word from register
.P
.B I2CWW h r wv 
SMB Write Word Data: write word to register
.P
.B I2CRK h r 
SMB Read Block Data: read data from register
.P
.B I2CWK h r bvs 
SMB Write Block Data: write data to register
.P
.B I2CWI h r bvs 
SMB Write I2C Block Data
.P
.B I2CRI h r num 
SMB Read I2C Block Data: read bytes from register
.P
.B I2CRD h num 
I2C read device
.P
.B I2CWD h bvs 
I2C write device
.P
.B I2CPC h r wv 
SMB Process Call: exchange register with word
.P
.B I2CPK h r bvs 
SMB Block Process Call: exchange data bytes with register
.P
.B I2CZ h bvs 
I2C zip
.P
.SS NOTIFICATIONS
.B NO  
Notification open
.P
.B NC h 
Notification close
.P
.B NP h 
Notification pause
.P
.B NR h 
Notification resume
.P
.SS SCRIPTS
.B PROC t 
Script store
.P
.B PROCR h pars 
Script run
.P
.B PROCU h pars 
Script update parameters
.P
.B PROCP h 
Script get status and parameters
.P
.B PROCS h 
Script stop
.P
.B PROCD h 
Script delete
.P
.B PARSE t 
Script validate
.P
.SS SERIAL
.B SERO dev b sef 
Serial open device
.P
.B SERC h 
Serial close device
.P
.B SERRB  
Serial read byte
.P
.B SERWB h bv 
Serial write byte
.P
.B SERR h num 
Serial read bytes
.P
.B SERW h bvs 
Serial write bytes
.P
.B SERDA h 
Serial data available
.P
.SS SHELL
.B SHELL name str 
Execute a shell command
.P
.SS SPI
.B SPIO spd spc b spf 
SPI open device
.P
.B SPIC h 
SPI close device
.P
.B SPIR h num 
SPI read bytes
.P
.B SPIW h bvs 
SPI write bytes
.P
.B SPIX h bvs 
SPI transfer bytes
.P
.SS UTILITIES
.B LGV  
Get lg library version
.P
.B SBC  
Get SBC's host name
.P
.B CGI cid 
Get internal configuration setting
.P
.B CSI cid v 
Set internal configuration setting
.P
.B T/TICK  
Get nanoseconds since the epoch
.P
.B MICS v 
Microseconds delay
.P
.B MILS v 
Milliseconds delay
.P
.B U/USER  
Set user
.P
.B C/SHARE  
Set share
.P
.B LCFG  
Reload permits configuration file
.P
.B PCD  
Print daemon configuration directory
.P
.B PWD  
Print daemon working directory
.P

.SH COMMANDS

.br
.SS FILES
.br

.IP "\fBFO file mode\fP - File open"
.IP "" 4
This is a privileged command.  See \fBpermits\fP.

.br
This function returns a handle to a file opened in a specified mode.

.br
Upon success a handle (>=0) is returned.  On error a negative status code
will be returned.

.br
The mode may have the following values.

.br

.EX
        Value   Meaning
READ      1     open file for reading
WRITE     2     open file for writing
RW        3     open file for reading and writing

.EE

.br
The following values may be or'd into the mode.

.br

.EX
         Value   Meaning
APPEND     4     All writes append data to the end of the file
CREATE     8     The file is created if it doesn't exist
TRUNC     16     The file is truncated

.EE

.br
Newly created files are owned by the user that launched the daemon
with permissions owner read and write.

.br

\fBExample\fP
.br

.EX
ls /ram/*.c
.br
/ram/q.c     /ram/qdhtxx.c  /ram/q-errcod.c  /ram/q_t1.c
.br
/ram/q-c1.c  /ram/Q-err.c   /ram/q-group.c   /ram/q_t2.c
.br

.br
$ rgs c 1 fo /ram/q.c 1  # read access
.br
1
.br

.br
$ rgs c 1 fo /ram/new.c 1 # file does not exist
.br
-58
.br
ERROR: file open failed
.br

.br
$rgs c 1 fo /ram/new.c 9 # can not create file
.br
-67
.br
ERROR: no permission to access file
.br

.EE

.br

.IP "\fBFC h\fP - File close"
.IP "" 4
This command closes a file previously opened by \fBFO\fP.

.br
Upon success nothing is returned.  On error a negative status code
will be returned.

.br

\fBExample\fP
.br

.EX
$ rgs c 1 fc 1 # First close okay.
.br

.br
$ rgs c 1 fc 1 # Second fails.
.br
-5
.br
ERROR: unknown handle
.br

.EE

.br

.br

.IP "\fBFR h num\fP - File read"
.IP "" 4
This command returns up to \fBnum\fP bytes of data read from the file.

.br
Upon success the count of returned bytes followed by the bytes themselves
is returned.  On error a negative status code will be returned.

.br

\fBExample\fP
.br

.EX
$ rgs c 1 fr 0 10
.br
5 48 49 128 144 255
.br

.br
$ rgs c 1 fr 0 10
.br
0
.br

.EE

.br

.IP "\fBFW h bvs\fP - File write"
.IP "" 4
This command writes \fBbvs\fP bytes to the file.

.br
Upon success nothing is returned.  On error a negative status code
will be returned.

.br

\fBExample\fP
.br

.EX
$ rgs c 1 fw 0 23 45 67 89
.br

.EE

.br

.IP "\fBFS h num from\fP - File seek"
.IP "" 4
This command seeks to a position within the file.

.br
The number of bytes to move is \fBnum\fP.  Positive offsets
move forward, negative offsets backwards.  The move start
position is determined by \fBfrom\fP as follows.

.br

.EX
    From
0   start
1   current position
2   end

.EE

.br
Upon success the new byte position within the file (>=0) is
returned.  On error a negative status code will be returned.

.br

\fBExample\fP
.br

.EX
$ rgs c 1 fs 0 200 0 # Seek to start of file plus 200
.br
200
.br

.br
$ rgs c 1 fs 0 0 1 # Return current position
.br
200
.br

.br
$ rgs c 1 fs 0 0 2 # Seek to end of file, return size
.br
296235
.br

.EE

.br

.IP "\fBFL pat num\fP - File list"
.IP "" 4
This command returns a list of the files matching \fBpat\fP. Up
to \fBnum\fP bytes may be returned.

.br
Upon success the count of returned bytes followed by the matching
files is returned.  On error a negative status code will be returned.

.br
A newline (0x0a) character separates each file name.

.br
This is a privileged command.  See \fBpermits\fP.

.br

\fBExample\fP
.br

.EX
$ rgs -a fl "/sys/bus/w1/devices/28*/w1_slave" 5000
.br
90 /sys/bus/w1/devices/28-000005d34cd2/w1_slave
.br
/sys/bus/w1/devices/28-001414abbeff/w1_slave
.br

.br
$ rgs -a fl "/sys/bus/*" 5000
.br
ERROR: no permission to access file
.br
-67
.br

.EE

.br
.SS GPIO
.br

.IP "\fBGO gc\fP - gpiochip open device"
.IP "" 4

.br
This is a privileged command.  See \fBpermits\fP.

.br
This command opens a gpiochip.

.br

\fBExample\fP
.br

.EX
$ rgs c 1 go 0 # open /dev/gpiochip0
.br
1
.br
$ rgs c 1 go 23 # try to open /dev/gpiochip23
.br
-78
.br
ERROR: can not open gpiochip
.br

.EE

.br

.IP "\fBGC h\fP - gpiochip close device"
.IP "" 4

.br
This command closes a gpiochip previously opened by \fBGO\fP.

.br

\fBExample\fP
.br

.EX
$ rgs c 1 gc 1 # first close ok
.br
$ rgs c 1 gc 1 # already closed
.br
-5
.br
ERROR: unknown handle
.br

.EE

.br

.IP "\fBGIC h\fP - gpiochip information"
.IP "" 4

.br
This command gets information for an opened gpiochip.  In particular
it gets the number of GPIO on the gpiochip, its name, and its usage.

.br

\fBExample\fP
.br

.EX
$ rgs c 1 gic 1
.br
54 "gpiochip0" "pinctrl-bcm2835"
.br

.EE

.br

.IP "\fBGIL h g\fP - gpiochip line information"
.IP "" 4

.br
This command gets information for GPIO \fBg\fP of an opened gpiochip.
In particular it gets the GPIO number, line flags, its user,
and its purpose.

.br
The meaning of the line flags bits are as given for the mode by \fBGMODE\fP.

.br
The user and purpose fields are filled in by the software which has
claimed the GPIO and may be blank.

.br

\fBExample\fP
.br

.EX
$ for ((i=2; i<10; i++)); do rgs c 1 gil 1 $i; done
.br
2 65536 "SDA1" ""
.br
3 65536 "SCL1" ""
.br
4 65536 "GPIO_GCLK" ""
.br
5 65536 "GPIO5" ""
.br
6 65536 "GPIO6" ""
.br
7 7 "SPI_CE1_N" "spi0 CS1"
.br
8 7 "SPI_CE0_N" "spi0 CS0"
.br
9 65536 "SPI_MISO" ""
.br

.EE

.br

.IP "\fBGMODE h g\fP - GPIO get mode"
.IP "" 4

.br
This command gets the mode for GPIO \fBg\fP of an opened gpiochip.

.br

.EX
Bit   Value   Meaning
0      1      Kernel: In use by the kernel
1      2      Kernel: Output
2      4      Kernel: Active low
3      8      Kernel: Open drain
4     16      Kernel: Open source
5     32      Kernel: Pull up set
6     64      Kernel: Pull down set
7     128     Kernel: Pulls off set
8     256     LG: Input
9     512     LG: Output
10    1024    LG: Alert
11    2048    LG: Group
12    4096    LG: ---
13    8192    LG: ---
14    16384   LG: ---
15    32768   LG: ---
16    65536   Kernel: Input
17    1<<17   Kernel: Rising edge alert
18    1<<18   Kernel: Falling edge alert
19    1<<19   Kernel: Realtime clock alert

.EE

.br
The LG bits are only set if the query was made by the process that
owns the GPIO.

.br

.IP "\fBGSI h g\fP - GPIO claim for input (simple)"
.IP "" 4

.br
This command claims GPIO \fBg\fP for input.

.br

\fBExample\fP
.br

.EX
$ rgs c 1 gsi 1 23 # claim GPIO 23 for input.
.br

.EE

.br

.IP "\fBGSIX h lf g\fP - GPIO claim for input"
.IP "" 4

.br
This command claims GPIO \fBg\fP for input.

.br
The line flags \fBlf\fP may be used to set the GPIO
as active low, open drain, open source,
pull up, pull down, pull off.

.br

\fBExample\fP
.br

.EX
$ rgs c 1 gsi 1 0 23 # claim GPIO 23 for input.
.br

.EE

.br

.IP "\fBGSO h g\fP - GPIO claim for output (simple)"
.IP "" 4

.br

rgs.c
/*
This is free and unencumbered software released into the public domain.

Anyone is free to copy, modify, publish, use, compile, sell, or
distribute this software, either in source code form or as a compiled
binary, for any purpose, commercial or non-commercial, and by any
means.

In jurisdictions that recognize copyright laws, the author or authors
of this software dedicate any and all copyright interest in the
software to the public domain. We make this dedication for the benefit
of the public at large and to the detriment of our heirs and
successors. We intend this dedication to be an overt act of
relinquishment in perpetuity of all present and future rights to this
software under copyright law.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

For more information, please refer to <http://unlicense.org/>
*/

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <time.h>
#include <inttypes.h>
#include <ctype.h>
#include <unistd.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <netdb.h>
#include <arpa/inet.h>

#include "lgpio.h"
#include "rgpiod.h"

#include "lgCmd.h"
#include "lgDbg.h"
#include "lgMD5.h"

/*
This program provides a socket interface to some of
the commands available from lg.
*/

#define RGS_VERSION 0x00020000

#define RGS_CONNECT_ERR 255
#define RGS_OPTION_ERR  254
#define RGS_SCRIPT_ERR  253

char text[CMD_MAX_EXTENSION];

int printFlags = 0;

int status = LG_OKAY;

#define SOCKET_OPEN_FAILED -1

#define PRINT_HEX 1
#define PRINT_ASCII 2

static char *xCmdUsage = "\n\
C                 Set share\n\
CGI cid           Get internal configuration setting\n\
CSI cid v         Set internal configuration setting\n\
\n\
FC h              File close\n\
FL pat num        File list\n\
FO file mode      File open\n\
FR h num          File read\n\
FS h num from     File seek\n\
FW h bvs          File write\n\
\n\
GBUSY h g k       GPIO or group tx busy\n\
GC h              gpiochip close device\n\
GDEB h g us       GPIO debounce time\n\
GGR h g           GPIO group read\n\
GGW h g gbits     GPIO group write (simple)\n\
GGWX h g gbits gmask  | GPIO group write\n\
GIC h             gpiochip information\n\
GIL h g           gpiochip line information\n\
GO gc             gpiochip open device\n\
GP h g mon moff   GPIO tx pulses (simple)\n\
GPX h g mon moff off cyc  | GPIO tx pulses\n\
GR h g            GPIO read\n\
GROOM h g k       GPIO or group tx entries\n\
GSA h g           GPIO claim for alerts (simple)\n\
GSAX h lf ef g nfyh  | GPIO claim for alerts\n\
GSF h g           GPIO free\n\
GSGF h g          GPIO group free\n\
GSGI h g*         GPIO group claim for inputs (simple)\n\
GSGIX h lf g*     GPIO group claim for inputs\n\
GSGO h g*         GPIO group claim for outputs (simple)\n\
GSGOX h lf g* v*  GPIO group claim for outputs\n\
GSI h g           GPIO claim for input (simple)\n\
GSIX h lf g       GPIO claim for input\n\
GSO h g           GPIO claim for output\n\
GSOX h lf g v     GPIO claim for output\n\
GW h g v          GPIO write\n\
GWAVE h g p*      GPIO group tx wave\n\
GWDOG h g us      GPIO watchdog time\n\
\n\
I2CC h            I2C close device\n\
I2CO ib id if     I2C open device\n\
I2CPC h r wv      SMB Process Call: exchange register with word\n\
I2CPK h r bvs     SMB Block Process Call: exchange data bytes with register\n\
I2CRB h r         SMB Read Byte Data: read byte from register\n\
I2CRD h num       I2C read device\n\
I2CRI h r num     SMB Read I2C Block Data: read bytes from register\n\
I2CRK h r         SMB Read Block Data: read data from register\n\
I2CRS h           SMB Read Byte: read byte\n\
I2CRW h r         SMB Read Word Data: read word from register\n\
I2CWB h r bv      SMB Write Byte Data: write byte to register\n\
I2CWD h bvs       I2C write device\n\
I2CWI h r bvs     SMB Write I2C Block Data\n\
I2CWK h r bvs     SMB Write Block Data: write data to register\n\
I2CWQ h bit       SMB Write Quick: write bit\n\
I2CWS h bv        SMB Write Byte: write byte\n\
I2CWW h r wv      SMB Write Word Data: write word to register\n\
I2CZ h bvs        I2C zip\n\
\n\
LCFG              Reload permits configuration file\n\
LGV               Get lg library version\n\
\n\
MICS v            Microseconds delay\n\
MILS v            Milliseconds delay\n\
\n\
NC h              Notification close\n\
NO                Notification open\n\
NP h              Notification pause\n\
NR h              Notification resume\n\
\n\
P h g pf pdc      GPIO tx PWM (simple)\n\
PARSE t           Script validate\n\
PCD               Print daemon configuration directory\n\
PROC t            Script store\n\
PROCD h           Script delete\n\
PROCP h           Script get status and parameters\n\
PROCR h pars      Script run\n\
PROCS h           Script stop\n\
PROCU h pars      Script update parameters\n\
PWD               Print daemon working directory\n\
PX h g pf pdc off cyc  | GPIO tx PWM\n\
\n\
S h g spw         GPIO tx servo pulses (simple)\n\
SBC               Get SBC's host name\n\
SERC h            Serial close device\n\
SERDA h           Serial data available\n\
SERO dev b sef    Serial open device\n\
SERR h num        Serial read bytes\n\
SERRB h           Serial read byte\n\
SERW h bvs        Serial write bytes\n\
SERWB h bv        Serial write byte\n\
SHELL name str    Execute a shell command\n\
SPIC h            SPI close device\n\
SPIO spd spc b spf  | SPI open device\n\
SPIR h num        SPI read bytes\n\
SPIW h bvs        SPI write bytes\n\
SPIX h bvs        SPI transfer bytes\n\
SX h g sf spw off cyc  | GPIO tx servo pulses\n\
SHARE             Set share\n\
\n\
T                 Get nanoseconds since the epoch\n\
TICK              Get nanoseconds since the epoch\n\
\n\
U                 Set user\n\
USER              Set user\n\
\n\
Numbers may be entered as hex (prefix 0x), octal (prefix 0),\n\
otherwise they are assumed to be decimal.\n\
\n\
Examples\n\
\n\
rgs u test1 s 1 i2co 1 0x20 0 # get handle to device 0x20 on I2C bus 1\n\
\n\
man rgs for full details.\n\
\n";

static uint64_t xMakeSalt(void)
{
   struct timespec xts;

   clock_gettime(CLOCK_REALTIME, &xts);

   return ((xts.tv_sec + xts.tv_nsec) * random()) + (random() + xts.tv_nsec);
}

static void xReport(int err, char *fmt, ...)
{
   char buf[128];
   va_list ap;

   if (err > status) status = err;

   va_start(ap, fmt);
   vsnprintf(buf, sizeof(buf), fmt, ap);
   va_end(ap);

   fprintf(stderr, "%s\n", buf);

   fflush(stderr);
}

static int xInitOpts(int argc, char *argv[])
{
   int opt, args;

   opterr = 0;

   args = 1;

   while ((opt = getopt(argc, argv, "ahvx")) != -1)
   {
      switch (opt)
      {
         case 'a':
            printFlags |= PRINT_ASCII;
            args++;
            break;

         case 'h':
            printf("%s", xCmdUsage);
            args++;
            break;

         case 'x':
            printFlags |= PRINT_HEX;
            args++;
            break;

         case 'v':
            printf("rgs_%d.%d.%d.%d\n",
               (RGS_VERSION>>24)&0xff, (RGS_VERSION>>16)&0xff,
               (RGS_VERSION>>8)&0xff, RGS_VERSION&0xff);
            args++;
            break;

         default:
            args++;
            xReport(RGS_OPTION_ERR, "ERROR: bad option %c", optopt);
      }
    }
   return args;
}

static int xOpenSocket(void)
{
   int sock, err;
   struct addrinfo hints, *res, *rp;
   const char *addrStr, *portStr;

   portStr = getenv(LG_ENVPORT);

   if (!portStr) portStr = LG_DEFAULT_SOCKET_PORT_STR;

   addrStr = getenv(LG_ENVADDR);

   if (!addrStr) addrStr = LG_DEFAULT_SOCKET_ADDR_STR;

   memset (&hints, 0, sizeof (hints));

   hints.ai_family   = PF_UNSPEC;
   hints.ai_socktype = SOCK_STREAM;
   hints.ai_flags   |= AI_CANONNAME;

   err = getaddrinfo(addrStr, portStr, &hints, &res);

   if (err) return SOCKET_OPEN_FAILED;

   for (rp=res; rp!=NULL; rp=rp->ai_next)
   {
      sock = socket(rp->ai_family, rp->ai_socktype, rp->ai_protocol);

      if (sock == -1) continue;

      if (connect(sock, rp->ai_addr, rp->ai_addrlen) != -1) break;
   }

   freeaddrinfo(res);

   if (rp == NULL) return SOCKET_OPEN_FAILED;

   return sock;
}

static void xShowResult(int rv, lgCmd_p cmdP, char *cmdExt)
{
   int i, r, ch;
   uint32_t *argI=(uint32_t*)&cmdP[1];
   uint64_t *argQ=(uint64_t*)&cmdP[1];

   r = cmdP->status;

   switch (rv)
   {
      case 0:
      case 1:
         if (r < 0)
         {
            printf("%d\n", r);
            xReport(RGS_SCRIPT_ERR, "ERROR: %s", lguErrorText(r));
         }
         break;

      case 2:
         printf("%d\n", r);
         if (r < 0) xReport(RGS_SCRIPT_ERR, "ERROR: %s", lguErrorText(r));
         break;

      case 3:
         if (r < 0)
         {
            printf("%d\n", r);
            xReport(RGS_SCRIPT_ERR, "ERROR: %s", lguErrorText(r));
         }
         else printf("%"PRIu64"\n", argQ[0]);
         break;

      case 5:
         printf("lg_%d.%d.%d.%d\n",
            (cmdP->status>>24)&0xff, (cmdP->status>>16)&0xff,
            (cmdP->status>>8)&0xff, cmdP->status&0xff);
         break;

      case 6: /*
                 FL  FR  I2CPK  I2CRD  I2CRI  I2CRK
                 I2CZ  PCD PWD SBC SERR  SPIX  SPIR USER
              */

         if ((cmdP->cmd == LG_CMD_PCD) ||
             (cmdP->cmd == LG_CMD_PWD) ||
             (cmdP->cmd == LG_CMD_SBC))
         {
            printf("%s\n", cmdExt);
            break;
         }

         printf("%d", r);
         if (r < 0) xReport(RGS_SCRIPT_ERR, "ERROR: %s", lguErrorText(r));
         if (r > 0)
         {
            if (printFlags == PRINT_ASCII) printf(" ");

            for (i=0; i<r; i++)
            {
               ch = cmdExt[i];

               if (printFlags & PRINT_HEX) printf(" %hhx", ch);

               else if (printFlags & PRINT_ASCII)
               {
                  if (isprint(ch) || (ch == '\n') || (ch == '\r'))
                     printf("%c", ch);
                  else printf("\\x%02hhx", ch);
               }
               else printf(" %hhu", cmdExt[i]);
            }
         }
         printf("\n");
         break;

      case 7: /* PROCP */
         if (r != (4 + (4*LG_MAX_SCRIPT_PARAMS)))
         {
            printf("%d", r);
            xReport(RGS_SCRIPT_ERR, "ERROR: %s", lguErrorText(r));
         }
         else
         {
            printf("%d", argI[0]);
            for (i=0; i<LG_MAX_SCRIPT_PARAMS; i++)
            {
               printf(" %d", argI[i+1]);
            }
         }
         printf("\n");
         break;

      case 9: /* GGR */
         if (r < 0)
         {
            printf("%d\n", r);
            xReport(RGS_SCRIPT_ERR, "ERROR: %s", lguErrorText(r));
         }
         else
         {
            printf("%d %"PRIu64"\n", argI[2], argQ[0]);
         }
         break;

      case 10: /* GIC */
         if (r < 0)
         {
            printf("%d\n", r);
            xReport(RGS_SCRIPT_ERR, "ERROR: %s", lguErrorText(r));
         }
         else
         {
            printf("%d \"%s\" \"%s\"\n", argI[0], cmdExt+4, cmdExt+36);
         }
         break;

      case 11: /* GIL */
         if (r < 0)
         {
            printf("%d\n", r);
            xReport(RGS_SCRIPT_ERR, "ERROR: %s", lguErrorText(r));
         }
         else
         {
            printf("%d %d \"%s\" \"%s\"\n",
               argI[0], argI[1], cmdExt+8, cmdExt+40);
         }
         break;

      default:
         printf("*** command=%d, status=%d\n", cmdP->cmd, r);
         if (r < 0) xReport(RGS_SCRIPT_ERR, "ERROR: %s", lguErrorText(r));
         break;

   }
}

static int xSendCommand(int sock, lgCmd_p cmdP, char *cmdExt)
{
   int msgLen;

   if (sock == SOCKET_OPEN_FAILED)
   {
      xReport(RGS_CONNECT_ERR, "socket connect failed");
      return -1;
   }

   msgLen = sizeof(lgCmd_t) + cmdP->size;

   /*
   printf("cmd=%s\n", lgDbgStr2Hex(sizeof(lgCmd_t), (char *)cmdP));
   */

   if (send(sock, cmdP, msgLen, 0) != msgLen)
   {
      xReport(RGS_CONNECT_ERR, "socket send failed");
      return -1;
   }

   if (recv(sock, cmdP, sizeof(lgCmd_t), MSG_WAITALL) != sizeof(lgCmd_t))
   {
      xReport(RGS_CONNECT_ERR, "socket receive failed");
      return -1;
   }

   if (cmdP->size)
   {
      if (recv(sock, cmdExt, cmdP->size, MSG_WAITALL) != cmdP->size)
      {
         xReport(RGS_CONNECT_ERR, "socket receive failed");
         return -1;
      }
      cmdExt[cmdP->size] = 0;
   }
   return 0;
}

int main(int argc , char *argv[])
{
   int sock;
   int args, idx, i, pp, l, len;
   char salt1[LG_SALT_LEN];
   char user[LG_USER_LEN];
   const char *userStr, *shareStr;
   cmdCtl_t ctl;
   cmdScript_t s;
   lgCmd_t cmdBuf[CMD_MAX_EXTENSION/sizeof(lgCmd_t)];
   lgCmd_p cmdP=cmdBuf;
   char *cmdExt=(char*)&cmdP[1];

   sock = xOpenSocket();

   args = xInitOpts(argc, argv);

   text[0] = 0;
   l = 0;
   pp = 0;

   userStr = getenv(LG_ENVUSER);

   if (userStr && strlen(userStr))
   {
      sprintf(text, "u %s ", userStr);
      l = strlen(text);
      pp = l;
   }

   shareStr = getenv(LG_ENVSHARE);

   if (shareStr && strlen(shareStr))
   {
      sprintf(text+pp, "c %s ", shareStr);
      l = strlen(text);
      pp = l;
   }

   for (i=args; i<argc; i++)
   {
      l += (strlen(argv[i]) + 1);
      if (l < sizeof(text)) {sprintf(text+pp, "%s ", argv[i]); pp=l;}
   }

   if (pp) {text[--pp] = 0;}

   ctl.inScript = 0;
   ctl.eaten = 0;

   len = strlen(text);
   idx = 0;

   while ((idx >= 0) && (ctl.eaten < len))
   {
      if ((idx=cmdParse(text, &ctl, cmdBuf, sizeof(cmdBuf))) >= 0)
      {
         cmdP->magic = LG_MAGIC;
         cmdP->doubles = 0;
         cmdP->longs = 0;
         cmdP->shorts = 0;

         if (cmdP->cmd < LG_CMD_SCRIPT)
         {
            if (cmdP->cmd == LG_CMD_PARSE)
            {
               cmdParseScript(cmdExt, &s, 1);
               if (s.par) free (s.par);
            }
            else if (cmdP->cmd == LG_CMD_USER)
            {
               /* need to auto login, create salt */
               snprintf(salt1, LG_SALT_LEN, "%015"PRIx64, xMakeSalt());
               /* save username */
               snprintf(user, LG_USER_LEN, "%s", cmdExt);
               /* change command to salt + user */
               sprintf(cmdExt, "%s.%s", salt1, user);
               cmdP->size = strlen(cmdExt);
               
               if (xSendCommand(sock, cmdP, cmdExt) == 0)
               {
                  /* take salt2 from message and overwrite with hash */
                  lgMd5UserHash(user, salt1, cmdExt, "", cmdExt);
                  cmdP->cmd = LG_CMD_PASSW;
                  cmdP->size = strlen(cmdExt);
                  if (xSendCommand(sock, cmdP, cmdExt) == 0)
                     xShowResult(0, cmdP, cmdExt);
               }
            }
            else
            {
               if (xSendCommand(sock, cmdP, cmdExt) == 0)
                  xShowResult(cmdInfo[idx].rv, cmdP, cmdExt);
            }
         }
         else xReport(RGS_SCRIPT_ERR,
                 "%s only allowed within a script", cmdInfo[idx].name);
      }
      else
      {
         if (idx == CMD_UNKNOWN_CMD)
            xReport(RGS_SCRIPT_ERR,
               "%s? unknown command, rgs -h for help", cmdStr());
         else
            xReport(RGS_SCRIPT_ERR,
               "%s: bad parameter, rgs -h for help", cmdStr());
      }
   }

   if (sock >= 0) close(sock);

   return status;
}

